File: src/instruction.rs
//! Virtual Machine Instructions
//!
//! An Instruction contains information about a single instruction such as the
//! type and arguments.

/// Enum containing all possible instruction types.
#[derive(Debug, Clone)]
#[repr(u16)]
pub enum InstructionType {
    SetInteger               = 0,
    SetFloat                 = 1,
    SetString                = 2,
    SetObject                = 3,
    SetArray                 = 4,
    SetName                  = 5,
    GetIntegerPrototype      = 6,
    GetFloatPrototype        = 7,
    GetStringPrototype       = 8,
    GetArrayPrototype        = 9,
    GetThreadPrototype       = 10,
    GetTruePrototype         = 11,
    GetFalsePrototype        = 12,
    GetMethodPrototype       = 13,
    GetCompiledCodePrototype = 14,
    SetTrue                  = 15,
    SetFalse                 = 16,
    SetLocal                 = 17,
    GetLocal                 = 18,
    SetConst                 = 19,
    GetConst                 = 20,
    SetAttr                  = 21,
    GetAttr                  = 22,
    SetCompiledCode          = 23,
    Send                     = 24,
    Return                   = 25,
    GotoIfFalse              = 26,
    GotoIfTrue               = 27,
    Goto                     = 28,
    DefMethod                = 29,
    DefLiteralMethod         = 30,
    RunCode                  = 31,
    GetToplevel              = 32,
    IsError                  = 33,
    ErrorToString            = 34,
    IntegerAdd               = 35,
    IntegerDiv               = 36,
    IntegerMul               = 37,
    IntegerSub               = 38,
    IntegerMod               = 39,
    IntegerToFloat           = 40,
    IntegerToString          = 41,
    IntegerBitwiseAnd        = 42,
    IntegerBitwiseOr         = 43,
    IntegerBitwiseXor        = 44,
    IntegerShiftLeft         = 45,
    IntegerShiftRight        = 46,
    IntegerSmaller           = 47,
    IntegerGreater           = 48,
    IntegerEquals            = 49,
    StartThread              = 50,
    FloatAdd                 = 51,
    FloatMul                 = 52,
    FloatDiv                 = 53,
    FloatSub                 = 54,
    FloatMod                 = 55,
    FloatToInteger           = 56,
    FloatToString            = 57,
    FloatSmaller             = 58,
    FloatGreater             = 59,
    FloatEquals              = 60,
    ArrayInsert              = 61,
    ArrayAt                  = 62,
    ArrayRemove              = 63,
    ArrayLength              = 64,
    ArrayClear               = 65,
    StringToLower            = 66,
    StringToUpper            = 67,
    StringEquals             = 68,
    StringToBytes            = 69,
    StringFromBytes          = 70,
    StringLength             = 71,
    StringSize               = 72,
    StdoutWrite              = 73,
    StderrWrite              = 74,
    StdinRead                = 75,
    StdinReadLine            = 76,
    FileOpen                 = 77,
    FileWrite                = 78,
    FileRead                 = 79,
    FileReadLine             = 80,
    FileFlush                = 81,
    FileSize                 = 82,
    FileSeek                 = 83,
    RunFileFast              = 84
}

/// Struct for storing information about a single instruction.
#[derive(Clone)]
pub struct Instruction {
    /// The type of instruction.
    pub instruction_type: InstructionType,

    /// The arguments of the instruction.
    pub arguments: Vec<usize>,

    /// The line from which the instruction originated.
    pub line: usize,

    /// The column from which the instruction originated.
    pub column: usize
}

impl Instruction {
    /// Returns a new Instruction.
    pub fn new(ins_type: InstructionType, arguments: Vec<usize>, line: usize,
               column: usize) -> Instruction {
        Instruction {
            instruction_type: ins_type,
            arguments: arguments,
            line: line,
            column: column
        }
    }

    pub fn arg(&self, index: usize) -> Result<usize, String> {
        self.arguments
            .get(index)
            .cloned()
            .ok_or(format!("undefined instruction argument {}", index))
    }
}

File: compiler/bootstrap/lib/inko/instruction.rb
module Inko
  class Instruction
    attr_reader :name, :arguments, :line, :column

    NAME_MAPPING = [
      :set_integer,
      :set_float,
      :set_string,
      :set_object,
      :set_array,
      :get_integer_prototype,
      :get_float_prototype,
      :get_string_prototype,
      :get_array_prototype,
      :get_true_prototype,
      :get_false_prototype,
      :get_method_prototype,
      :get_compiled_code_prototype,
      :get_true,
      :get_false,
      :set_local,
      :get_local,
      :set_literal_const,
      :get_literal_const,
      :set_literal_attr,
      :get_literal_attr,
      :set_compiled_code,
      :send_literal,
      :return,
      :goto_if_false,
      :goto_if_true,
      :goto,
      :def_method,
      :def_literal_method,
      :run_code,
      :get_toplevel,
      :is_error,
      :integer_add,
      :integer_div,
      :integer_mul,
      :integer_sub,
      :integer_mod,
      :integer_to_float,
      :integer_to_string,
      :integer_bitwise_and,
      :integer_bitwise_or,
      :integer_bitwise_xor,
      :integer_shift_left,
      :integer_shift_right,
      :integer_smaller,
      :integer_greater,
      :integer_equals,
      :spawn_literal_process,
      :float_add,
      :float_mul,
      :float_div,
      :float_sub,
      :float_mod,
      :float_to_integer,
      :float_to_string,
      :float_smaller,
      :float_greater,
      :float_equals,
      :array_insert,
      :array_at,
      :array_remove,
      :array_length,
      :array_clear,
      :string_to_lower,
      :string_to_upper,
      :string_equals,
      :string_to_bytes,
      :string_from_bytes,
      :string_length,
      :string_size,
      :stdout_write,
      :stderr_write,
      :stdin_read,
      :stdin_read_line,
      :file_open,
      :file_write,
      :file_read,
      :file_read_line,
      :file_flush,
      :file_size,
      :file_seek,
      :run_literal_file,
      :run_file,
      :send,
      :get_self,
      :get_binding_prototype,
      :get_binding,
      :set_const,
      :get_const,
      :set_attr,
      :get_attr,
      :literal_const_exists,
      :run_literal_code,
      :set_prototype,
      :get_prototype,
      :local_exists,
      :get_caller,
      :literal_responds_to,
      :responds_to,
      :literal_attr_exists,
      :set_outer_scope,
      :spawn_process,
      :send_process_message,
      :receive_process_message,
      :get_current_pid,
      :set_parent_local,
      :get_parent_local,
      :get_binding_of_caller
    ].each_with_index.each_with_object({}) do |(value, index), hash|
      hash[value] = index
    end

    # Instructions where the register containing a value to return is the last
    # register, instead of the first one.
    REGISTER_LAST = [
      :set_const,
      :set_literal_const,
      :set_attr,
      :set_literal_attr,
      :set_local,
      :set_parent_local
    ]

    def initialize(name, arguments, line, column)
      @name = name
      @arguments = arguments
      @line = line
      @column = column

      if !line or !column
        raise ArgumentError, 'A line and column number are required'
      end

      @arguments.each do |arg|
        unless arg.is_a?(Fixnum)
          raise TypeError, "arguments must be Fixnums, not a #{arg.class}"
        end
      end
    end

    def name_integer
      NAME_MAPPING.fetch(name)
    end

    def remap_arguments
      @arguments.map! { |arg| yield arg }
    end

    def written_register
      if REGISTER_LAST.include?(name.to_sym)
        arguments[-1]
      else
        arguments[0]
      end
    end

    def inspect
      "Instruction(#{name}, #{arguments.inspect}, line: #{line}, column: #{column})"
    end
  end
end
